File: commonroad_reach/utility/sweep_line.py
from collections import defaultdict
from enum import Enum, auto
from functools import cmp_to_key
from typing import Dict, List, Tuple, Union

from commonroad_reach.data_structure.reach.reach_line import ReachLine
from commonroad_reach.data_structure.reach.reach_polygon import ReachPolygon
from commonroad_reach.data_structure.segment_tree import CounterSegmentTree, ToggleSegmentTree


class EventType(Enum):
    """
    Possible types of events in the sweep line algorithm.
    """

    ENTER = auto()
    EXIT = auto()


class Event:
    """
    Event in the sweep line algorithm.

    p_lon = x (CART) = s (CVLN), p_lat = y (CART) = d (CVLN)

    It is assumed that the line is swept from left to right. The left edge of a rectangle
    is typed 'ENTER', and the right edge 'EXIT'.
    """

    def __init__(self, type_vent, p_lon, p_lat_low, p_lat_high):
        self.type = type_vent
        self.p_lon = p_lon
        self.p_lat_low = p_lat_low
        self.p_lat_high = p_lat_high

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Event):
            if self.type == other.type and self.p_lon == other.p_lon and \
                    self.p_lat_low == other.p_lat_low and self.p_lat_high == other.p_lat_high:
                return True

        return False

    def __repr__(self) -> str:
        return f"Event(type={self.type}, p_lon={self.p_lon}, " \
               f"p_lat_low={self.p_lat_low}, p_lat_high={self.p_lat_high})"


class SweepLine:
    """
    Class performing sweep line algorithms.
    """
    tree: Union[CounterSegmentTree, ToggleSegmentTree]

    @classmethod
    def obtain_vertical_segments_from_rectangles(cls, list_rectangles: List[ReachPolygon], ) -> List[ReachLine]:
        """
        Returns a list of vertical segments from the input rectangles.

        Steps:
            1. create a segment tree with min/max lateral position of rectangles
            2. create events of line sweeping from rectangles: left edge = ENTER, right edge = EXIT
            3. create vertical segments with the events
        """
        if not list_rectangles:
            return []

        p_lat_min_rectangles, p_lat_max_rectangles = \
            cls.compute_extremum_lateral_positions_of_rectangles(list_rectangles)
        cls.tree = CounterSegmentTree(p_lat_min_rectangles, p_lat_max_rectangles)

        list_events = cls.create_event_list(list_rectangles)

        list_segments_vertical = cls.create_vertical_segments_from_events(list_events)

        return list_segments_vertical

    @staticmethod
    def compute_extremum_lateral_positions_of_rectangles(list_rectangles: List[ReachPolygon]) -> Tuple[int, int]:
        """
        Returns the minimum and maximum lateral positions of the given list of rectangles.
        """
        p_lat_min_rectangles = min([rectangle.p_lat_min for rectangle in list_rectangles])
        p_lat_max_rectangles = max([rectangle.p_lat_max for rectangle in list_rectangles])

        return p_lat_min_rectangles, p_lat_max_rectangles

    @classmethod
    def create_event_list(cls, list_rectangles: List[ReachPolygon]) -> List[Event]:
        """
        Creates a list of sorted events with the given list of rectangles.
        """
        list_events = []

        for rectangle in list_rectangles:
            list_events.append(Event(EventType.ENTER, rectangle.p_lon_min,
                                     rectangle.p_lat_min, rectangle.p_lat_max))
            list_events.append(Event(EventType.EXIT, rectangle.p_lon_max,
                                     rectangle.p_lat_min, rectangle.p_lat_max))

        list_events = cls.sort_events(list_events)

        return list_events

    @classmethod
    def sort_events(cls, list_events: List[Event]) -> List[Event]:
        return sorted(list_events, key=cmp_to_key(cls.compare_events))

    @classmethod
    def compare_events(cls, event1: Event, event2: Event):
        """
        Custom comparison function for events.

        Events are ordered in the following order:
            1. longitudinal position of rectangles
            2. type of the event
            3. lower lateral position
        """
        if event1.p_lon < event2.p_lon:
            return -1

        elif event1.p_lon > event2.p_lon:
            return 1

        if event1.type == EventType.ENTER and event2.type == EventType.EXIT:
            return -1

        elif event1.type == EventType.EXIT and event2.type == EventType.ENTER:
            return 1

        if event1.p_lat_low < event2.p_lat_low:
            return -1

        elif event1.p_lat_low > event2.p_lat_low:
            return 1

        return 0

    @classmethod
    def create_vertical_segments_from_events(cls, list_events: List[Event]) -> List[ReachLine]:
        """
        Creates a list of vertical segments from the list of events.
        """
        list_segments_vertical = []

        for event in list_events:
            if event.type == EventType.ENTER:
                list_segments_vertical += cls.create_vertical_segments_from_event(event)
                cls.tree.activate(event.p_lat_low, event.p_lat_high)

            else:
                cls.tree.deactivate(event.p_lat_low, event.p_lat_high)
                list_segments_vertical += cls.create_vertical_segments_from_event(event)

        return list_segments_vertical

    @classmethod
    def create_vertical_segments_from_event(cls, event: Event) -> List[ReachLine]:
        """
        Returns a list of vertical segments with the tree and event.

        For each event, query the tree to get the nonactive intervals, which is the desired vertical segment.
        """
        list_segments = []

        stack_intervals_lat = cls.tree.get_non_active_intervals(event.p_lat_low, event.p_lat_high)
        while stack_intervals_lat:
            p_lat_high = stack_intervals_lat.pop()
            p_lat_low = stack_intervals_lat.pop()
            p_lon = event.p_lon

            segment = ReachLine(p_lon, p_lat_low, p_lon, p_lat_high)
            list_segments.append(segment)

        return list_segments

    @classmethod
    def create_rectangles_from_vertical_segments(cls, list_segments: List[ReachLine]) -> List[ReachPolygon]:
        """
        Returns a list of rectangles from the given vertical segments.
            
        Step:
            1. Create a segment tree with the list of segments.
            2. Create a dictionary that maps p_lon to a list of rectangles whose left edge is aligned with p_lon.
            3. Merge rectangles that share the same coordinates of p_lat.
        """
        cls.tree = cls.create_tree_from_segments(list_segments)

        dict_p_lon_to_list_rectangles = cls.create_p_lon_to_rectangles_dictionary(list_segments)

        list_rectangles_final = cls.merge_rectangles_with_same_lateral_coordinates(dict_p_lon_to_list_rectangles)

        return list_rectangles_final

    @staticmethod
    def create_tree_from_segments(list_segments: List[ReachLine]) -> ToggleSegmentTree:
        """
        Creates a ToggleSegmentTree from the list of given segments.
        """
        p_lat_min_segments = min([segment.p_lat_min for segment in list_segments])
        p_lat_max_segments = max([segment.p_lat_max for segment in list_segments])

        tree = ToggleSegmentTree(int(p_lat_min_segments), int(p_lat_max_segments))

        return tree

    @classmethod
    def create_p_lon_to_rectangles_dictionary(cls, list_segments: List[ReachLine]) -> Dict[int, List[ReachPolygon]]:
        """
        Create a dictionary that maps p_lon to a list of rectangles whose left edge is aligned with p_lon.

        Steps:
            1. Create a dictionary that maps p_lon to list of tuples of p_lat from segments
            2. Iterate through p_lon, retrieve relevant tuples of p_lat and toggle the status of segments between
               these p_lat in the segment tree. Get intervals of active segments and create rectangles.
        """
        if not list_segments:
            return {}

        dict_p_lon_to_list_rectangles = defaultdict(list)

        dict_p_lon_to_list_tuples_p_lat = defaultdict(list)
        for segment in list_segments:
            p_lon_min, p_lat_min, _, p_lat_max = segment.bounds
            dict_p_lon_to_list_tuples_p_lat[p_lon_min].append((p_lat_min, p_lat_max))

        list_p_lon = list(dict_p_lon_to_list_tuples_p_lat.keys())
        for p_lon_min, p_lon_max in zip(list_p_lon[:-1], list_p_lon[1:]):
            list_tuples_p_lat = dict_p_lon_to_list_tuples_p_lat[p_lon_min]
            for tuple_p_lat in list_tuples_p_lat:
                cls.tree.toggle(tuple_p_lat[0], tuple_p_lat[1])

            stack_interval_active = cls.tree.get_stack_of_active_intervals()
            while stack_interval_active:
                p_lat_max = stack_interval_active.pop()
                p_lat_min = stack_interval_active.pop()

                dict_p_lon_to_list_rectangles[p_lon_min].append(
                    ReachPolygon.from_rectangle_vertices(p_lon_min, p_lat_min, p_lon_max, p_lat_max))

        return dict_p_lon_to_list_rectangles

    @classmethod
    def merge_rectangles_with_same_lateral_coordinates(
            cls, dict_p_lon_to_list_rectangles: Dict[int, List[ReachPolygon]]) -> List[ReachPolygon]:
        """
        Return a list of rectangles with possible merging.

        Iterate through pairs of lists of rectangles, if there is a right rectangle with the same lateral coordinates,
        then do not add to list. Instead, the right rectangle is popped and replaced by the merged one.
        """
        list_rectangles_merged = []

        list_p_lon = list(dict_p_lon_to_list_rectangles)
        for p_lon_left, p_lon_right in zip(list_p_lon[:-1], list_p_lon[1:]):
            list_rectangles_left = dict_p_lon_to_list_rectangles[p_lon_left]
            list_rectangles_right = dict_p_lon_to_list_rectangles[p_lon_right]

            for rectangle_left in list_rectangles_left:
                add_to_list = True

                for rectangle_right in list_rectangles_right:
                    if cls.rectangles_have_same_p_lat(rectangle_left, rectangle_right) and \
                            rectangle_left.p_lon_max >= rectangle_right.p_lon_min:
                        list_rectangles_right.remove(rectangle_right)
                        list_rectangles_right.append(
                            ReachPolygon.from_rectangle_vertices(rectangle_left.p_lon_min, rectangle_left.p_lat_min,
                                                                 rectangle_right.p_lon_max, rectangle_left.p_lat_max))
                        add_to_list = False
                        break

                if add_to_list:
                    list_rectangles_merged.append(rectangle_left)

        # add rectangles from the last list in to final list
        list_rectangles_merged += dict_p_lon_to_list_rectangles[list_p_lon[-1]]

        return list_rectangles_merged

    @staticmethod
    def rectangles_have_same_p_lat(rectangle1: ReachPolygon, rectangle2: ReachPolygon) -> bool:
        """
        Returns True if the two input rectangles have the same lateral positions.
        """
        return rectangle1.p_lat_min == rectangle2.p_lat_min and rectangle1.p_lat_max == rectangle2.p_lat_max

File: cpp/src/utility/sweep_line.cpp
#include "reachset/utility/sweep_line.hpp"
#include "reachset/utility/shared_using.hpp"

using namespace reach;

SweepLine::Event::Event(EventType const& event_type, double const& p_lon,
                        double const& p_lat_low, double const& p_lat_high) :
        type(event_type), p_lon(p_lon),
        p_lat_low(p_lat_low), p_lat_high(p_lat_high) {}

bool SweepLine::Event::operator==(const Event& other) const {
    if (this->type == other.type and
        this->p_lon == other.p_lon and
        this->p_lat_low == other.p_lat_low and
        this->p_lat_high == other.p_lat_high)
        return true;

    return false;
}

/// *Steps*:
/// 1. create a segment tree with min/max lateral position of rectangles
/// 2. create events of line sweeping: left = ENTER, right = EXIT
/// 3. create vertical segments with events
vector<ReachLinePtr>
SweepLine::obtain_vertical_segments_from_rectangles(vector<ReachPolygonPtr> const& vec_rectangles) {
    if (vec_rectangles.empty())
        return {};

    // Step 1
    auto[p_lat_min_rectangles, p_lat_max_rectangles] = compute_extremum_lateral_positions_of_rectangles(vec_rectangles);
    SweepLine::tree_counter = CounterSegmentTree(p_lat_min_rectangles, p_lat_max_rectangles);

    // Step 2
    auto vec_events = SweepLine::create_events(vec_rectangles);

    // Step 3
    auto vec_segments_vertical = create_vertical_segments_from_events(vec_events);

    return vec_segments_vertical;
}

tuple<double, double>
SweepLine::compute_extremum_lateral_positions_of_rectangles(vector<ReachPolygonPtr> const& vec_rectangles) {
    auto p_lat_min_rectangles = std::numeric_limits<double>::infinity();
    auto p_lat_max_rectangles = -std::numeric_limits<double>::infinity();

    for (auto const& rectangle: vec_rectangles) {
        p_lat_min_rectangles = std::min(p_lat_min_rectangles, rectangle->p_lat_min());
        p_lat_max_rectangles = std::max(p_lat_max_rectangles, rectangle->p_lat_max());
    }

    return {p_lat_min_rectangles, p_lat_max_rectangles};
}

vector<EventPtr> SweepLine::create_events(vector<ReachPolygonPtr> const& vec_rectangles) {
    vector<EventPtr> vec_events;
    vec_events.reserve(vec_rectangles.size());

    for (auto const& rectangle: vec_rectangles) {
        vec_events.emplace_back(make_shared<Event>(EventType::ENTER,
                                                   rectangle->p_lon_min(),
                                                   rectangle->p_lat_min(),
                                                   rectangle->p_lat_max()));

        vec_events.emplace_back(make_shared<Event>(EventType::EXIT,
                                                   rectangle->p_lon_max(),
                                                   rectangle->p_lat_min(),
                                                   rectangle->p_lat_max()));
    }

    sort_events(vec_events);

    return vec_events;
}

/// Comparisons are made in sequence on:
/// 1. longitudinal position of rectangles
/// 2. type of the event
/// 3. lower lateral position
void SweepLine::sort_events(std::vector<EventPtr>& vec_events) {
    std::sort(vec_events.begin(), vec_events.end(), [](EventPtr const& event_a, EventPtr const& event_b) {
        if (event_a->p_lon < event_b->p_lon)
            return true;
        else if (event_a->p_lon > event_b->p_lon)
            return false;

        if (event_a->type == EventType::ENTER and event_b->type == EventType::EXIT)
            return true;
        else if (event_a->type == EventType::EXIT and event_b->type == EventType::ENTER)
            return false;

        if (event_a->p_lat_low < event_b->p_lat_low)
            return true;
        else if (event_a->p_lat_low > event_b->p_lat_low)
            return false;

        return false;
    });
}

/// For each event, query the tree to get the nonactive intervals, which is the desired vertical segment.
vector<ReachLinePtr> SweepLine::create_vertical_segments_from_events(vector<EventPtr> const& vec_events) {
    vector<ReachLinePtr> vec_segments_vertical;

    for (auto const& event: vec_events) {
        if (event->type == EventType::ENTER) {
            auto vec_segments_from_event = create_vertical_segments_from_event(event);
            vec_segments_vertical.insert(vec_segments_vertical.end(),
                                         std::make_move_iterator(vec_segments_from_event.begin()),
                                         std::make_move_iterator(vec_segments_from_event.end()));

            tree_counter.activate(event->p_lat_low, event->p_lat_high);
        } else {
            tree_counter.deactivate(event->p_lat_low, event->p_lat_high);

            auto vec_segments_from_event = create_vertical_segments_from_event(event);
            vec_segments_vertical.insert(vec_segments_vertical.end(),
                                         std::make_move_iterator(vec_segments_from_event.begin()),
                                         std::make_move_iterator(vec_segments_from_event.end()));
        }
    }
    return vec_segments_vertical;
}

vector<ReachLinePtr> SweepLine::create_vertical_segments_from_event(EventPtr const& event) {
    vector<ReachLinePtr> vec_segments;

    auto stack_intervals_lat = SweepLine::tree_counter.get_stack_of_non_active_intervals(event->p_lat_low,
                                                                                         event->p_lat_high);

    while (not stack_intervals_lat.empty()) {
        auto p_lat_high = stack_intervals_lat.top();
        stack_intervals_lat.pop();
        auto p_lat_low = stack_intervals_lat.top();
        stack_intervals_lat.pop();
        auto p_lon = event->p_lon;

        vec_segments.emplace_back(make_shared<ReachLine>(p_lon, p_lat_low, p_lon, p_lat_high));
    }

    return vec_segments;
}

/// *Steps*:
/// 1. Create a segment tree with the list of segments.
/// 2. Create a dictionary that maps p_lon to a list of rectangles whose left edge is aligned with p_lon.
/// 3. Merge rectangles that share the same coordinates of of p_lat.
vector<ReachPolygonPtr>
SweepLine::create_rectangles_from_vertical_segments(vector<ReachLinePtr> const& vec_segments) {
    // Step 1
    auto[p_lat_min_segments, p_lat_max_segments] = compute_extremum_lateral_positions_of_segments(vec_segments);
    SweepLine::tree_toggle = ToggleSegmentTree(p_lat_min_segments, p_lat_max_segments);

    // Step 2
    auto map_p_lon_to_vec_rectangles = create_p_lon_to_rectangles_map(vec_segments);

    // Step 3
    auto vec_rectangles_final = merge_rectangles_with_same_lateral_coordinates(map_p_lon_to_vec_rectangles);

    return vec_rectangles_final;
}

tuple<double, double>
SweepLine::compute_extremum_lateral_positions_of_segments(vector<ReachLinePtr> const& vec_segments) {
    auto p_lat_min_segments = std::numeric_limits<double>::infinity();
    auto p_lat_max_segments = -std::numeric_limits<double>::infinity();

    for (auto const& segment: vec_segments) {
        p_lat_min_segments = std::min(p_lat_min_segments, segment->p_lat_min);
        p_lat_max_segments = std::max(p_lat_max_segments, segment->p_lat_max);
    }

    return {p_lat_min_segments, p_lat_max_segments};
}

/// *Steps*:
/// 1. Create a dictionary that maps p_lon to list of tuples of p_lat from segments
/// 2. Iterate through p_lon, retrieve relevant tuples of p_lat and toggle the status of segments between these p_lat
/// in the segment tree. Get intervals of active segments and create rectangles.
map<double, vector<ReachPolygonPtr>>
SweepLine::create_p_lon_to_rectangles_map(vector<ReachLinePtr> const& vec_segments) {
    if (vec_segments.empty())
        return {};

    map<double, vector<ReachPolygonPtr>> map_p_lon_to_vec_rectangles;

    // Step 1
    map<double, vector<tuple<double, double>>> map_p_lon_to_vec_tuples_p_lat;
    vector<double> vec_p_lon;
    vec_p_lon.reserve(vec_segments.size());

    for (auto const& segment: vec_segments) {
        map_p_lon_to_vec_tuples_p_lat[segment->p_lon_min].emplace_back(
                make_tuple(segment->p_lat_min, segment->p_lat_max));
        vec_p_lon.emplace_back(segment->p_lon_min);
    }
    vec_p_lon.erase(unique(vec_p_lon.begin(), vec_p_lon.end()), vec_p_lon.end());

    // Step 2
    for (int i = 0; i < vec_p_lon.size() - 1; ++i) {
        auto p_lon_min = vec_p_lon[i];
        auto p_lon_max = vec_p_lon[i + 1];
        auto vec_tuples_p_lat = map_p_lon_to_vec_tuples_p_lat[p_lon_min];

        for (auto const&[p_lat_min, p_lat_max]: vec_tuples_p_lat)
            tree_toggle.toggle(p_lat_min, p_lat_max);

        auto stack_intervals_active = tree_toggle.get_stack_of_active_intervals();
        while (not stack_intervals_active.empty()) {
            auto p_lat_max = stack_intervals_active.top();
            stack_intervals_active.pop();
            auto p_lat_min = stack_intervals_active.top();
            stack_intervals_active.pop();

            map_p_lon_to_vec_rectangles[p_lon_min].emplace_back(
                    make_shared<ReachPolygon>(p_lon_min, p_lat_min, p_lon_max, p_lat_max));
        }

    }

    return map_p_lon_to_vec_rectangles;
}

/// Iterate through pairs of lists of rectangles, if there is a right rectangle with the same lateral coordinates,
/// then do not add to list. Instead, the right rectangle is popped and replaced by the merged one.
vector<ReachPolygonPtr> SweepLine::merge_rectangles_with_same_lateral_coordinates(
        map<double, vector<ReachPolygonPtr>>& map_p_lon_to_vec_rectangles) {

    vector<ReachPolygonPtr> vec_rectangles_merged;

    vector<double> vec_p_lon;
    vec_p_lon.reserve(map_p_lon_to_vec_rectangles.size());
    for (auto const&[key, val]: map_p_lon_to_vec_rectangles) {
        vec_p_lon.emplace_back(key);
    }

    for (int i = 0; i < vec_p_lon.size() - 1; ++i) {
        auto p_lon_left = vec_p_lon[i];
        auto p_lon_right = vec_p_lon[i + 1];

        auto& vec_rectangles_left = map_p_lon_to_vec_rectangles[p_lon_left];
        auto& vec_rectangles_right = map_p_lon_to_vec_rectangles[p_lon_right];

        for (auto const& rectangle_left: vec_rectangles_left) {
            auto add_to_list = true;

            for (auto const& rectangle_right: vec_rectangles_right) {
                if (rectangles_have_same_p_lat(rectangle_left, rectangle_right) &&
                    rectangle_left->p_lon_max() >= rectangle_right->p_lon_min()) {
                    auto it_end = std::remove(vec_rectangles_right.begin(), vec_rectangles_right.end(),
                                              rectangle_right);
                    vec_rectangles_right.erase(it_end, vec_rectangles_right.end());

                    vec_rectangles_right.emplace_back(make_shared<ReachPolygon>(rectangle_left->p_lon_min(),
                                                                                rectangle_left->p_lat_min(),
                                                                                rectangle_right->p_lon_max(),
                                                                                rectangle_left->p_lat_max()));
                    add_to_list = false;
                    break;
                }
            }

            if (add_to_list)
                vec_rectangles_merged.emplace_back(rectangle_left);
        }
    }
    // add rectangles from the last vector into final vector
    for (auto const& rectangle: map_p_lon_to_vec_rectangles[vec_p_lon.back()]) {
        vec_rectangles_merged.emplace_back(rectangle);
    }

    return vec_rectangles_merged;
}

bool SweepLine::rectangles_have_same_p_lat(ReachPolygonPtr const& rectangle1, ReachPolygonPtr const& rectangle2) {
    return rectangle1->p_lat_min() == rectangle2->p_lat_min() and rectangle1->p_lat_max() == rectangle2->p_lat_max();
}